rustdoc: include external files in documentation (RFC 1990)

QuietMisdreavus opened the pull request:
Part of https://github.com/rust-lang/rfcs/pull/1990 (needs work on the error reporting, which i'm deferring to after this initial PR)

cc #44732 

Also fixes #42760, because the prep work for the error reporting made it easy to fix that at the same time.

rust-highfive commented:
r? @GuillaumeGomez

(rust_highfive has picked a reviewer for you, use r? to override)

QuietMisdreavus commented:
Travis failure is because my test for 42760 failed `-_-` Back to the drawing board...

QuietMisdreavus commented:
update: apparently `is_sugared_doc` is `false` for doc comments when you get them in `with_desugared_doc`. The issue was *not* fixed, because there's no way for it to see the difference between a doc comment and a raw doc attribute where it's extracting them. `>:(`

QuietMisdreavus commented:
So i wound up modifying `syntax::ast::Attribute::with_desugared_doc` to set `is_sugared_doc` on the attribute when it calls the given closure. I'm not sure if this is entirely valid w.r.t. the original intent of that method, but rustdoc was the only in-tree user of that method that i found, so it may work out anyway. It fixed the issue in the test rendering i had, so i'll let travis see if i messed anything else up.

QuietMisdreavus commented:
Travis was green, so i'm declaring this Ready To Be Looked At. I updated one test to make sure combining included docs and inline docs worked and squashed the commits. The libsyntax change is still in its own commit, though, since it was only included to make the fix for #42760 work properly, and i'm still not sure whether it's the totally right thing to do.

GuillaumeGomez commented on src/librustdoc/clean/mod.rs:
```diff
@@ -628,6 +747,15 @@ impl Attributes {
+                            match Attributes::load_include(include_path, &filename) {
+                                Ok(frag) => {
+                                    let line = doc_line;
+                                    doc_line += frag.len();
```
Oh so you decided to put in the "real" line number instead of the one from the original file?

QuietMisdreavus replied:
Yeah. My thought was that if you're trying to trace a line number from the total markdown blob, then you could use that to get the doc fragment, since each attribute is joined with a line break anyway.

GuillaumeGomez commented:
It'd be nice to have some outputs (or even tests?) for the case of failing doc codes.

QuietMisdreavus commented:
> It'd be nice to have some outputs (or even tests?) for the case of failing doc codes.

*That's* what i was forgetting! Give me a moment, i can make a quick run.

QuietMisdreavus commented:
With the following files:

<details>

`lib.rs`:

```rust
#![feature(external_doc)]

/// hey yo check out these docs, i can splice them in from another file
///
/// ```rust
/// panic!("oh no");
/// ```
///
#[doc(include = "some-file.md")]
pub struct SomeStruct;

/// Item docs.
///
#[doc="Hello there!"]
///
/// # Example
///
/// ```rust
/// // some code here
/// panic!("oh no");
/// ```
pub struct OtherStruct;

#[doc(include = "bad-docs.md")]
#[doc(include = "no-docs.md")]
pub struct AllTheStruct;
```

`some-file.md`:

```text
hey check out this included file

this is so sick, i can just paste this anywhere
```

`bad-docs.md`:

    # some bad docs

    ```rust
    panic!("oh no");
    ```

(`no-docs.md` deliberately missing)

</details>

Running `cargo test` gives the following output:

<details>

![image](https://user-images.githubusercontent.com/5217170/30778032-ab64e4c4-a08f-11e7-8763-81bc744cf90f.png)

</details>

This is with the latest force push. (I neglected to ask rustdoctest to combine doc fragments, so that's fixed now.)

GuillaumeGomez commented:
And a column as well for the "real" line number (so a `Span`?). If you want, I can work on that part (the test one).

QuietMisdreavus commented:
@GuillaumeGomez I added a `Span` to each `DocFragment` and made sure that they get properly combined in `collapse-docs`, does this seem all right to you?

GuillaumeGomez commented:
Please post a test failure output. :)

QuietMisdreavus commented:
I haven't touched the failure reporting code, so it'll look the same as in https://github.com/rust-lang/rust/pull/44781#issuecomment-331677063. Do you want me to make that part of this PR, or would you like to merge this as-is and take it up yourself afterward?

GuillaumeGomez commented:
I don't mind merging as is. Just tell me what you prefer to do.

QuietMisdreavus commented:
If i remember right, we discussed fixing up the error reporting in a separate PR, so that's what i was assuming we would do here (and my preference, just so we can get this in `>_>`). If you want to wrap it all into one PR, just tell me what code is likely to need the span information and i can take a shot at it.

QuietMisdreavus commented:
@ollie27 Oh yikes, that's a good point. On the other hand, i don't really see a good way to reliably do it cross-crate without asking rustc to load the file in. In the RFC discussion we talked about making it an alias for `#[doc=include_str!("file.md")]` which would do exactly that, but would require work within the compiler to make sure certain macros could be used within attributes and expand correctly. (That's also outside my expertise. `>_>`)

SimonSapin commented:
Having a special case for doc include in rustc might be easier than supporting arbitrary macros in arbitrary attributes. (Or maybe not, I don’t know.)

bors commented:
:umbrella: The latest upstream changes (presumably #45046) made this pull request unmergeable. Please resolve the merge conflicts.

aidanhs commented:
@QuietMisdreavus I've set this back as with author for the merge conflicts. After that, I'm also not clear after that if you're going to do more work or it's going back to review by @GuillaumeGomez?

QuietMisdreavus commented:
@aidanhs I'm actually not sure either. I'm concerned it may need to be scrapped entirely in favor of the `#[doc=include_str!()]` method alluded to above. I'm not entirely sure how to go about that, though - it would require work on the compiler frontend to extend https://github.com/rust-lang/rust/issues/34981 to include (at least) some builtin macros as an allowed expression in attributes, then also adding `#[doc(include="file.md")]` as sugar for that. Should we pull in compiler-team to take a look?

GuillaumeGomez commented:
Won't kill to add them in the loop. :)

cc @rust-lang/compiler

alexcrichton commented:
triage ping for @QuietMisdreavus, do you know what the next steps here are?

QuietMisdreavus commented:
@alexcrichton Same as last week; not really. As much as i'd like to push to get this imperfect implementation merged, it also seems kinda foolish if the "proper" implementation would wipe out part of it. I'm not sure who to pull in to push for the aforementioned `#[doc=include_str!("file.md")]` implementation, and since https://github.com/rust-lang/rust/issues/34981 is still outstanding i don't know how many things need to happen to get there.

Do i need to close this in lieu of more implementation planning?

alexcrichton commented:
I wouldn't personally bank on `include_str!` in an attribute ever working, but for whether or not to close this I personally prefer to close things that still need planning (aka no clear path forward), but it doesn't particularly matter either way.

QuietMisdreavus commented:
I've nominated this for the next @rust-lang/dev-tools triage meeting, to see what they say about the commentary in this thread.

A little history of this thread, since it's been open for a while:

- This is an implementation of `#[doc(include = "file.md")]`.
  - Right now the PR only lays groundwork for properly matching failing doctests to the proper lines, without updating the error code to use it.
- The current implementation handles file loading within rustdoc, loading the file in when the crate is being parsed post-compilation.
  - [This is a problem when an item is re-exported](https://github.com/rust-lang/rust/pull/44781#issuecomment-333196280), since the file path won't be relative to the right crate root any more. It *may* be possible to wrangle rustdoc into pulling from the right crate, *if* sources are available for it, by putting in some work to recognize something as a re-export. I haven't investigated this.
- This does *not* include any work toward making `#[doc=include_str!("file.md")]` work, like was discussed in the thread for this RFC. I'm... not actually sure what would be needed to get that to be recognized. `>_>`

I'm basically looking for advice on what to do with this PR because if i need to rework how files are pulled in i'd rather scrap it and start from scratch doing that. Otherwise, i'm stalled.

kennytm commented:
Hi @rust-lang/dev-tools 🙂 what's the status of the is PR, any advices that this could move forward?

QuietMisdreavus commented:
We talked about this PR in the dev-tools meeting this week. We decided it would be better to pursue working the file-loading (or at least some kind of path expansion) into early-expansion instead of within rustdoc, to try to get around the re-export issue noted earlier. I need to get in touch with someone on the compiler team to get a sense of how feasible this is.

QuietMisdreavus commented:
*/me kicks the PR*

With generous help from @eddyb, i now proclaim this PR resurrected! The major change from before was how the `include_path` is handled. Previously, both documentation runs and test runs would just grab the filename being passed into `rustdoc` and pass around its containing directory. Now, each run is handled differently:

* Test runs still calculate the `include_path` the same way, since only items documented inside the crate are tested. No need to worry about cross-crate imports if their documentation isn't getting pulled in the first place!
* Documentation runs, on the other hand, treat things a little differently. Instead of having a singular path in the `DocContext`, it now keeps track of a `current_crate` that it will use to poke a `TyCtxt` and `CodeMap` for the crate-root filename. This is effectively the same thing for local items, but now when something is inlined (specifically via `src/librustdoc/clean/inline.rs`) it will swap that `CrateNum` with the one for that item while it extracts doc comments. This way, it can go find the location on the filesystem where that other crate lives, and fetch the included docs from there.

To make sure cross-crate docs work with included docs, i've added another test to that effect.

ollie27 commented:
So what happens if the source for the external crate isn't available to `rustdoc`? For example `std` is distributed as `rlib`s. It really seems to me that `rustc` should be reading the files just like `include_str` and not `rustdoc`.

eddyb commented:
I thought the inlining was mainly used for the `libstd` facade but I'm probably wrong.

QuietMisdreavus commented:
How often do we expect to see crates that won't have their source available when downstream documentation is being made? Alternately, do we expect libstd to make heavy use of doc(include) after this lands?

As for inlining, it'll happen any time one crate wants to re-export something from another one. For example `reqwest` re-exports lots of stuff from `hyper`, and those docs are fully inlined.

I also think it would be more robust to have rustc itself load the files in, but this was the solution offered when i asked about it. Otherwise, i'm back to where i started, and i'd need pointers as to where it could actually get done.

ollie27 commented:
> How often do we expect to see crates that won't have their source available when downstream documentation is being made?

That's probably pretty unlikely but it isn't a requirement right now so it would be a shame if it became one.

eddyb commented:
If you need to include the actual text in the source, macro expansion should be fine... `cfg_attr`, if it's not too special, could serve as a guide. Alternatively, wherever `derive` is handled would work too.

QuietMisdreavus commented:
*Round 3, FIGHT!*

So now, with the help of @jseyfried, `libsyntax` now turns `#[doc(include="filename')]` into `#[doc(include(file="filename", contents="file contents"))]` at macro expansion time. Rustdoc takes this expanded form and turns that into a `DocFragment::Include` with the whole file's contents. (This means the broken-up lines of the previous iteration are now gone, since the expansion was made to just load the whole thing into one String all at once.) Everything else was able to work as i had it before, and the tests i previously wrote worked when i tried them locally.

I wound up squashing the commits up to clean the history, but i kept the `with_desugared_doc` change on its own since it was already separated and didn't need to be combined.

QuietMisdreavus commented:
...oh. It looks like the place where I did the expansion is before features get checked? Guess I need to check my attributes before i convert them.

```
[00:51:15] ---- [compile-fail] compile-fail/feature-gate-external_doc.rs stdout ----
[00:51:15] 	
[00:51:15] error: /checkout/src/test/compile-fail/feature-gate-external_doc.rs:11: unexpected error: '11:1: 11:26: couldn't read /checkout/src/test/compile-fail/asdf.md: No such file or directory (os error 2)'
[00:51:15] 
[00:51:15] error: /checkout/src/test/compile-fail/feature-gate-external_doc.rs:11: expected error not found: #[doc(include = "...")] is experimental
```

QuietMisdreavus commented:
Okay, with this last force-push it should properly detect the attribute before expanding it, and avoid attempting to load the file if the feature gate was not set.

shepmaster commented:
Ping from triage, @GuillaumeGomez — will you have time to review this soon?

GuillaumeGomez commented:
I'll require @QuietMisdreavus's assistance but I have time now.

bors commented:
:umbrella: The latest upstream changes (presumably #45039) made this pull request unmergeable. Please resolve the merge conflicts.

QuietMisdreavus commented:
I've rebased to resolve the merge conflict.

GuillaumeGomez commented on src/librustdoc/clean/mod.rs:
```diff
@@ -596,6 +662,47 @@ impl Attributes {
+        mi.meta_item_list().and_then(|list| {
+            for meta in list {
+                if meta.check_name("include") {
+                    //the actual compiled `#[doc(include="filename")]` gets expanded to
```
Please put a whitespace after `//`.

GuillaumeGomez commented on src/librustdoc/html/render.rs:
```diff
@@ -1893,8 +1914,9 @@ fn render_assoc_const_value(item: &clean::Item) -> String {
-    if let Some(s) = item.doc_value() {
-        render_markdown(w, s, item.source.clone(), cx.render_type, prefix, &cx.shared)?;
+    if let Some(s) = cx.shared.maybe_collapsed_doc_value(item) {
+        trace!("Doc block: =====\n{}\n=====", s);
```
Shouldn't `debug!` be used instead? (Don't really know the good macro we're supposed to use in such cases...)

QuietMisdreavus replied:
Probably? I'm not that solid on what counts as what either, but i'll go ahead and switch it over.

GuillaumeGomez commented on src/librustdoc/passes/collapse_docs.rs:
```diff
@@ -26,15 +44,51 @@ impl fold::DocFolder for Collapser {
+                match curr_frag {
+                    DocFragment::SugaredDoc(_, _, ref mut doc_string)
+                        | DocFragment::RawDoc(_, _, ref mut doc_string) => {
+                            //add a newline for extra padding between segments
```
Whitespace after `//` please.

GuillaumeGomez commented on src/libsyntax/ext/expand.rs:
```diff
@@ -1063,6 +1071,83 @@ impl<'a, 'b> Folder for InvocationCollector<'a, 'b> {
+
+            for it in list {
+                if !it.check_name("include") {
+                    items.push(noop_fold_meta_list_item(it, self));
```
It's just bikeshedding at this point but you could add a `continue` here and it'd allow you to avoid having the `else` condition (and its indent).

QuietMisdreavus replied:
I like this idea. I'm a fan of this.

GuillaumeGomez commented on src/libsyntax/ext/expand.rs:
```diff
@@ -1063,6 +1071,83 @@ impl<'a, 'b> Folder for InvocationCollector<'a, 'b> {
+                        let err_count = self.cx.parse_sess.span_diagnostic.err_count();
+                        self.check_attribute(&at);
+                        if self.cx.parse_sess.span_diagnostic.err_count() > err_count {
+                            //avoid loading the file if they haven't enabled the feature
```
Whitepace.

GuillaumeGomez commented:
All good for me from code reading. Just a few nits and it's good to go imo.

QuietMisdreavus commented:
I've force-pushed to address the review comments.

GuillaumeGomez commented:
Then it's all good for me. r=me once CI is green.

QuietMisdreavus commented:
@bors r=GuillaumeGomez

bors commented:
:pushpin: Commit 52ee203 has been approved by `GuillaumeGomez`

<!-- @bors r=GuillaumeGomez 52ee203d656f7939de65fa837eea58230151e71e -->

bors commented:
:hourglass: Testing commit 52ee203d656f7939de65fa837eea58230151e71e with merge 0893ecc8955191f702513ff3a19d420f9650fae5...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/305585357?utm_source=github_status&utm_medium=notification)

QuietMisdreavus commented:
Failure... doesn't look legit? I'm not caught up on the known spurious failures.

```
ERROR:  Could not find a valid gem 'aws-sdk-resources' (= 2.10.88) in any repository
ERROR:  Possible alternatives: aws-sdk-resources
/home/travis/.rvm/rubies/ruby-2.2.7/lib/ruby/site_ruby/2.2.0/rubygems/core_ext/kernel_require.rb:59:in `require': cannot load such file -- aws-sdk (LoadError)
	from /home/travis/.rvm/rubies/ruby-2.2.7/lib/ruby/site_ruby/2.2.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/provider.rb:85:in `requires'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/provider/s3.rb:6:in `<class:S3>'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/provider/s3.rb:5:in `<class:Provider>'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/provider/s3.rb:4:in `<module:DPL>'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/provider/s3.rb:3:in `<top (required)>'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/provider.rb:59:in `const_get'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/provider.rb:59:in `block in new'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/cli.rb:41:in `fold'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/provider.rb:56:in `new'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/cli.rb:31:in `run'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/cli.rb:7:in `run'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/bin/dpl:5:in `<top (required)>'
	from /home/travis/.rvm/gems/ruby-2.2.7/bin/dpl:23:in `load'
	from /home/travis/.rvm/gems/ruby-2.2.7/bin/dpl:23:in `<main>'
```

kennytm commented:
@bors retry #44159 

Looks like that previous attempt to fix it doesn't work.

bors commented:
:hourglass: Testing commit 52ee203d656f7939de65fa837eea58230151e71e with merge 3755fe95556e5db39ebe5963d9171f3d3ea9511a...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5453), [status-travis](https://travis-ci.org/rust-lang/rust/builds/305718781?utm_source=github_status&utm_medium=notification)
Approved by: GuillaumeGomez
Pushing 3755fe95556e5db39ebe5963d9171f3d3ea9511a to master...
